runtime/gc_glue_java/MarkingDelegate.hpp

```diff
@@ -112,63 +119,79 @@ class MM_MarkingDelegate
-		case GC_ObjectModel::SCAN_ATOMIC_MARKABLE_REFERENCE_OBJECT:
-		case GC_ObjectModel::SCAN_MIXED_OBJECT:
-		case GC_ObjectModel::SCAN_OWNABLESYNCHRONIZER_OBJECT:
-		case GC_ObjectModel::SCAN_CLASS_OBJECT:
-		case GC_ObjectModel::SCAN_CLASSLOADER_OBJECT:
-			objectScanner = GC_MixedObjectScanner::newInstance(env, objectPtr, scannerSpace, 0);
-			*sizeToDo = sizeof(fomrobject_t) + ((GC_MixedObjectScanner *)objectScanner)->getBytesRemaining();
-			break;
-		case GC_ObjectModel::SCAN_POINTER_ARRAY_OBJECT:
-		{
-			uintptr_t slotsToDo = 0;
-			uintptr_t startIndex = setupPointerArrayScanner(env, objectPtr, reason, sizeToDo, &slotsToDo);
-			objectScanner = GC_PointerArrayObjectScanner::newInstance(env, objectPtr, scannerSpace, GC_ObjectScanner::indexableObject, slotsToDo, startIndex);
-			break;
-		}
-		case GC_ObjectModel::SCAN_FLATTENED_ARRAY_OBJECT:
-		{
-			Assert_MM_true(J9_IS_J9CLASS_FLATTENED(clazz));
-			Assert_MM_unimplemented();
-			break;
-		}
-		case GC_ObjectModel::SCAN_REFERENCE_MIXED_OBJECT:
-		{
-			fomrobject_t *referentSlotPtr = setupReferenceObjectScanner(env, objectPtr, reason);
-			objectScanner = GC_ReferenceObjectScanner::newInstance(env, objectPtr, referentSlotPtr, scannerSpace, 0);
-			*sizeToDo = sizeof(fomrobject_t) + ((GC_ReferenceObjectScanner *)objectScanner)->getBytesRemaining();
-			break;
-		}
-		case GC_ObjectModel::SCAN_PRIMITIVE_ARRAY_OBJECT:
-			/* nothing to do */
-			*sizeToDo = 0;
-			break;
-		default:
-			Assert_MM_unreachable();
-		}
+#if defined (J9VM_GC_REALTIME)
+		if (IS_ITEM_ARRAYLET((uintptr_t)objectPtr)) {
+			J9JavaVM* vm = (J9JavaVM*)_omrVM->_language_vm;
+			objectScanner = GC_PointerArrayObjectScanner::newInstance(env, objectPtr, ITEM_TO_ARRAYLET(objectPtr), scannerSpace, GC_ObjectScanner::indexableObjectNoSplit, (vm->arrayletLeafSize / sizeof(fj9object_t)), 0);
+		} else {
+#endif /* defined (J9VM_GC_REALTIME) */
+			J9Class *clazz = J9GC_J9OBJECT_CLAZZ(objectPtr, env);
+			/* object class must have proper eye catcher */
+			Assert_MM_true((UDATA)0x99669966 == clazz->eyecatcher);
+
+			switch(_extensions->objectModel.getScanType(objectPtr)) {
+			case GC_ObjectModel::SCAN_MIXED_OBJECT_LINKED:
+			case GC_ObjectModel::SCAN_ATOMIC_MARKABLE_REFERENCE_OBJECT:
+			case GC_ObjectModel::SCAN_MIXED_OBJECT:
+			case GC_ObjectModel::SCAN_OWNABLESYNCHRONIZER_OBJECT:
+			case GC_ObjectModel::SCAN_CLASS_OBJECT:
+			case GC_ObjectModel::SCAN_CLASSLOADER_OBJECT:
+				objectScanner = GC_MixedObjectScanner::newInstance(env, objectPtr, scannerSpace, 0);
+				*sizeToDo = sizeof(fomrobject_t) + ((GC_MixedObjectScanner *)objectScanner)->getBytesRemaining();
+				break;
+			case GC_ObjectModel::SCAN_POINTER_ARRAY_OBJECT:
+			{
+				uintptr_t slotsToDo = 0;
+				uintptr_t flags = 0;
+				fomrobject_t *basePtr = 0;
+				uintptr_t sizeInElements = 0;
+				uintptr_t startIndex = setupPointerArrayScanner(env, objectPtr, reason, sizeToDo, &slotsToDo, &sizeInElements, &basePtr, &flags);
```

Comment:
I'd slightly simplify the signature of setupPointerArrayScanner so that instead of passing triplet sizeToDo, &slotsToDo, &sizeInElements, pass only the last one (and call it sizeInElementsToDo), and calculate the first two here.